packages/notifi-react-card/lib/components/subscription/subscription-card-views/HistoryCardView.tsx

```diff
@@ -68,15 +70,84 @@ export const AlertHistoryView: React.FC<AlertHistoryViewProps> = ({
     setCardView({ state: 'preview' });
   };
 
-  const { client } = useNotifiClientContext();
+  const [endCursor, setEndCursor] = useState<string | undefined>();
+  const [hasNextPage, setHasNextPage] = useState<boolean | null>(null);
+  const [currentIndex, setCurrentIndex] = useState<number | null>();
+  const [visibleRange, setVisibleRange] = useState<ListRange>();
+  const [isScrolling, setIsScrolling] = useState<boolean | null>();
+  const [allNodes, setAllNodes] = useState<
+    ReadonlyArray<NotificationHistoryEntry>
+  >([]);
 
   const [alertHistoryData, setAlertHistoryData] =
     useState<GetNotificationHistoryResult>();
+
+  const { client } = useNotifiClientContext();
+
+  async function getNotificationHistory({
+    first,
+    after,
+  }: GetNotificationHistoryInput) {
+    const notificationHistory = await client
+      .getNotificationHistory({
+        first,
+        after,
+      })
+      .then((result) => {
+        setAlertHistoryData(result);
+        if (result.nodes) setAllNodes(allNodes.concat(result.nodes));
+        setEndCursor(result.pageInfo.endCursor);
+        setHasNextPage(result.pageInfo.hasNextPage);
+      });
+    return notificationHistory;
+  }
+
   useEffect(() => {
-    client.getNotificationHistory({}).then((result) => {
-      setAlertHistoryData(result);
-    });
-  }, [client]);
+    if (!alertHistoryData) {
+      getNotificationHistory({
+        first: MESSAGES_PER_PAGE,
+      });
+    }
+
+    const isRequestNextPage =
+      currentIndex &&
+      visibleRange &&
+      currentIndex === visibleRange?.endIndex &&
+      currentIndex > 0 &&
+      hasNextPage &&
+      endCursor &&
+      isScrolling;
+
+    if (isRequestNextPage) {
+      getNotificationHistory({
+        first: MESSAGES_PER_PAGE,
+        after: endCursor,
+      });
+    }
+  }, [currentIndex, visibleRange, hasNextPage, endCursor]);
+
+  const returnVirtualizedContainer = useCallback(() => {
+    if (!allNodes) return;
+    return (
+      <Virtuoso
+        style={{
+          height: notificationListHeight || '400px',
+          marginBottom: '25px',
+        }}
+        isScrolling={setIsScrolling}
+        rangeChanged={setVisibleRange}
+        data={allNodes.filter(
+          (notification) => notification.detail != undefined,
+        )}
+        itemContent={(index, notification) => {
+          setCurrentIndex(index);
+          return (
+            <AlertCard key={notification.id} notification={notification} />
+          );
+        }}
+      />
+    );
+  }, [alertHistoryData]);
```

Comment:
wouldn't this just be a component that takes setIsScrolling, setVisibleRange, setCurrentIndex, and allNodes as props?

it's tricky because the callback re-renders on alertHistoryData, but actually has a dependency on allNodes.

I think memoized render functions are a little harder to maintain, so my suggestion is to extract a separate component.

Reply:
Nice, catch! I will componetize the virtual container and update the dependency of allNodes (I must have been playing around with both as data at one point).

Reply:
I reviewed the code more closely and I realized that I will not need a useCallback for this because I am now appending the paginated node data into `allNodes`. 

I originally was trying to implement a `useCallback` as a workaround to render data at that index prior to having an array with all of the paginated data. I have reverted it back.

Here is the video of the testing to confirm that this works:

https://user-images.githubusercontent.com/108922885/198160510-fe6b4e71-93dc-4c96-9f52-ef0586ee2e97.mov

Please let me know if you have any other feedback!